sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -172,60 +172,47 @@ case class CreateMap(children: Seq[Expression]) extends Expression {
 }
 
 /**
- * Returns a Row containing the evaluation of all children expressions.
- */
-@ExpressionDescription(
-  usage = "_FUNC_(col1, col2, col3, ...) - Creates a struct with the given field values.")
-case class CreateStruct(children: Seq[Expression]) extends Expression {
-
-  override def foldable: Boolean = children.forall(_.foldable)
-
-  override lazy val dataType: StructType = {
-    val fields = children.zipWithIndex.map { case (child, idx) =>
+* a helper mixin to encapsulate [[CreateStruct]] => [[CreateNamedStruct]] transformation.
+*/
+sealed trait CreateStructLikeFactory[T <: Expression] extends (Seq[Expression] => T) {
+  protected case class CretaeStructHelper( children : Seq[Expression] ) {
+    private def attributeNames: Seq[String] = for {
+      (child, idx) <- children.zipWithIndex
+    } yield {
       child match {
-        case ne: NamedExpression =>
-          StructField(ne.name, ne.dataType, ne.nullable, ne.metadata)
-        case _ =>
-          StructField(s"col${idx + 1}", child.dataType, child.nullable, Metadata.empty)
+        case ne: NamedExpression => ne.name
+        case _ => s"col${idx + 1}"
       }
     }
-    StructType(fields)
-  }
-
-  override def nullable: Boolean = false
-
-  override def eval(input: InternalRow): Any = {
-    InternalRow(children.map(_.eval(input)): _*)
-  }
 
-  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    val rowClass = classOf[GenericInternalRow].getName
-    val values = ctx.freshName("values")
-    ctx.addMutableState("Object[]", values, s"this.$values = null;")
+    private def mkNamedStructArgs: Seq[Expression] = {
+      for {
```

Comment:
And if you want to write something that is even shorter (and less understandable):
```scala
attributeNames.map(Literal.apply).zip(children).flatMap(_.productIterator)
```